### PR TITLE
Pubsub flexible sample

### PR DESCRIPTION
**Pull request by jabubake**
Flex pubsub push endpoint sample with Datastore persistence.
Tests : TODO (given we're consolidating the flex apps for testing, removed the existing ones for now for this sample)

### CODE REVIEW

**Comment by lesv** [comment]
@jabubake Circle failed :(

File: flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubService.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+  private Publisher publisher;
+  private Logger logger = Logger.getLogger(this.getClass().getName());
+
+  /** Create a lazy initialized singleton of PubSubService */
+  private static class LazyInit {
+    private static final PubSubService INSTANCE;
+
+    static {
+      System.out.println("instance");
```
Comment (frankyn) [inline flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubService.java]: Is this line necessary?

File: flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubService.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+public class PubSubService {
+
+  private Publisher publisher;
+  private Logger logger = Logger.getLogger(this.getClass().getName());
+
+  /** Create a lazy initialized singleton of PubSubService */
+  private static class LazyInit {
+    private static final PubSubService INSTANCE;
+
+    static {
```
Comment (frankyn) [inline flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubService.java]: After reading into this Lazy initialization singleton pattern. I wonder if there's a more readable solution? This wasn't clear during my first pass.

File: flexible/pubsub/pom.xml [hunk 3]
```diff
@@ -0,0 +1,114 @@
+<!--
+  Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (frankyn) [inline flexible/pubsub/pom.xml]: New sample change year to 2016.

File: flexible/pubsub/src/main/java/com/example/flexible/pubsub/MessageRepository.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+   * Retrieve most recent messages.
+   *
+   * @param limit number of messages
+   * @return list of messages
+   */
+  public List<Message> retrieve(int limit) {
+    // Get Message saved in Datastore
+    Datastore datastore = getDatastoreInstance();
+    Query<Entity> query =
+        Query.newEntityQueryBuilder()
```
Comment (frankyn) [inline flexible/pubsub/src/main/java/com/example/flexible/pubsub/MessageRepository.java]: This should fit in the previous line.

File: flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubService.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+  /**
+   * Creates/modifies subscription with push configuration on topic
+   *
+   * @param topicId topic-id
+   * @param subscriptionId subscription-id
+   * @param endpoint push endpoint URL
+   * @throws Exception
+   */
+  private void addPushEndPoint(String topicId, String subscriptionId, String endpoint)
```
Comment (frankyn) [inline flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubService.java]: This method wasn't easy to read. Can you make this method be more descriptive?

File: flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubHome.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      sb.append("<tr>");
+      sb.append("<td>" + message.getMessageId() + "</td>");
+      sb.append("<td>" + message.getData() + "</td>");
+      sb.append("<td>" + message.getPublishTime() + "</td>");
+      sb.append("</tr>");
+    }
+    return sb.toString();
+  }
+
+  private PubSubHome() {
```
Comment (frankyn) [inline flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubHome.java]: I have a Bias to placing placing closing bracket on the same line as the open bracket. Like this: `private PubSubHome() { }`.

File: flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubHome.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+  private static MessageRepository messageRepository = MessageRepositoryImpl.getInstance();
+  private static int MAX_MESSAGES = 10;
+
+  /**
+   * Retrieve received messages in html.
+   *
+   * @return html representation of messages (one per row)
+   */
+  public static String getReceivedMessages() {
```
Comment (frankyn) [inline flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubHome.java]: Cleaner solution 👍

File: flexible/pubsub/src/main/java/com/example/flexible/pubsub/MessageRepository.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  void save(Message message);
+
+  /**
+   * Retrieve most recent stored messages.
+   * @param limit number of messages
+   * @return list of messages
+   */
+  List<Message> retrieve(int limit);
+}
+
```
Comment (frankyn) [inline flexible/pubsub/src/main/java/com/example/flexible/pubsub/MessageRepository.java]: Clean up whitespace

File: flexible/pubsub/pom.xml [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+      </plugin>
+      <plugin>
+        <groupId>org.eclipse.jetty</groupId>
+        <artifactId>jetty-maven-plugin</artifactId>
+        <version>${jetty.maven.plugin}</version>
+      </plugin>
+    </plugins>
+  </build>
+</project>
+<!-- [END project] -->
```
Comment (frankyn) [inline flexible/pubsub/pom.xml]: Add a new line.

File: flexible/pubsub/README.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+Copy the sample apps to your local machine, and cd to the pubsub directory:
+
+```
+git clone https://github.com/GoogleCloudPlatform/java-docs-samples
+cd java-docs-samples/flexible/pubsub
+```
+
+## Setup
+
+Make sure `gcloud` is installed and authenticated.
```
Comment (frankyn) [inline flexible/pubsub/README.md]: Add a link to Cloud SDK https://cloud.google.com/sdk/docs/

File: flexible/pubsub/README.md [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+Create a topic
+```
+gcloud beta pubsub topics create <your-topic-name>
+```
+
+Create a subscription, which includes specifying the endpoint to which the Pub/Sub server should send requests.
```
Comment (frankyn) [inline flexible/pubsub/README.md]: Maybe change this to: 

Create a [push subscription](https://cloud.google.com/pubsub/docs/push), to send pushed messages to a Google Cloud Project URL such as `https://<your-project-id>.appspot.com/`.

File: flexible/pubsub/README.md (hunk 10 above)
Comment (jabubake) [inline flexible/pubsub/README.md]: Done.

File: flexible/pubsub/README.md (hunk 11 above)
Comment (jabubake) [inline flexible/pubsub/README.md]: Done. That does sound better.

File: flexible/pubsub/README.md [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+### Send fake subscription push messages with:
+
+```
+curl -H "Content-Type: application/json" -i --data @sample_message.json
+"localhost:8080/pubsub/push?token=<your-token>"
+```
+
+## Deploy
+
+Update the environment variables `PUBSUB_TOPIC` and `PUBSUB_VERIFICATION_TOKEN` in `app.yaml`,
```
Comment (jabubake) [inline flexible/pubsub/README.md]: Done.

File: flexible/pubsub/src/main/webapp/index.jsp [hunk 13]
```diff
@@ -0,0 +1,21 @@
+<%@ page import="com.example.flexible.pubsub.PubSubHome" %>
```
Comment (jabubake) [inline flexible/pubsub/src/main/webapp/index.jsp]: Done.

File: flexible/pubsub/src/main/webapp/index.jsp [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    <table border="1" cellpadding="10">
+      <tr>
+      <th>Id</th>
+      <th>Data</th>
+      <th>PublishTime</th>
+      </tr>
+      <%= PubSubHome.getReceivedMessages() %>
+    </table>
+  </body>
+</html>
```
Comment (jabubake) [inline flexible/pubsub/src/main/webapp/index.jsp]: Done.

File: flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubPublish.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+      publisher.publish(pubsubMessage);
+      // redirect to home page
+      resp.sendRedirect("/");
+      // [END publish]
+    } catch (Exception e) {
+      resp.sendError(HttpStatus.SC_INTERNAL_SERVER_ERROR, e.getMessage());
+    }
+  }
+
+  // ...
```
Comment (jabubake) [inline flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubPublish.java]: I was trying to show that it was an incomplete class but it seems useless, removed.

File: flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubPublish.java [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  PubSubPublish(Publisher publisher) {
+    this.publisher = publisher;
+  }
+}
+
+
+
+
```
Comment (jabubake) [inline flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubPublish.java]: Thanks, done. also cleaned up the other files.

File: flexible/pubsub/pom.xml [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+
+  <parent>
+    <artifactId>doc-samples</artifactId>
+    <groupId>com.google.cloud</groupId>
+    <version>1.0.0</version>
+    <relativePath>../..</relativePath>
+  </parent>
+
+  <properties>
+    <appengine.maven.plugin>1.0.0</appengine.maven.plugin>
```
Comment (jabubake) [inline flexible/pubsub/pom.xml]: Good catch, done.

File: flexible/pubsub/pom.xml (hunk 9 above)
Comment (jabubake) [inline flexible/pubsub/pom.xml]: done

File: flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubHome.java (hunk 6 above)
Comment (jabubake) [inline flexible/pubsub/src/main/java/com/example/flexible/pubsub/PubSubHome.java]: Updated.
One gets spoiled when IntelliJ automatically does that for you on view.

File: flexible/pubsub/src/main/java/com/example/flexible/pubsub/MessageRepository.java (hunk 8 above)
Comment (jabubake) [inline flexible/pubsub/src/main/java/com/example/flexible/pubsub/MessageRepository.java]: Done.

**Pull request merged**